v1/remote/image.go

```diff
@@ -15,20 +15,150 @@
+		Host:   r.ref.Context().RegistryStr(),
+		Path:   fmt.Sprintf("/v2/%s/%s/%s", r.ref.Context().RepositoryStr(), resource, identifier),
+	}
+}
+
+func (r *remoteImage) MediaType() (types.MediaType, error) {
+	// TODO(jonjohnsonjr): Determine this based on response.
+	return types.DockerManifestSchema2, nil
+}
+
+// TODO(jonjohnsonjr): Handle manifest lists.
+// TODO(jonjohnsonjr): DockerHub returns the manifest list's digest when it falls back to schema 2??
+func (r *remoteImage) RawManifest() ([]byte, error) {
+	r.manifestLock.Lock()
+	defer r.manifestLock.Unlock()
+	if r.manifest != nil {
+		return r.manifest, nil
+	}
+
+	u := r.url("manifests", r.ref.Identifier())
+	req, err := http.NewRequest("GET", u.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Set("Accept", string(types.DockerManifestSchema2))
+	resp, err := r.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		// TODO(jonjohnsonjr): Parse this into a structured error.
+		return nil, fmt.Errorf("unrecognized status code during manifest GET: %v", resp.Status)
+	}
+
+	manifest, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	digest, _, err := v1.SHA256(ioutil.NopCloser(bytes.NewReader(manifest)))
+	if err != nil {
+		return nil, err
+	}
+
+	// Validate the digest matches what we asked for, if pulling by digest.
+	if dgst, ok := r.ref.(name.Digest); ok {
+		if digest.String() != dgst.DigestStr() {
+			return nil, fmt.Errorf("manifest digest: %s does not match requested digest: %s", digest, dgst.DigestStr())
+		}
+	} else if checksum := resp.Header.Get("Docker-Content-Digest"); checksum != "" && checksum != digest.String() {
+		// When pulling by tag, we can only validate that the digest matches what the registry told us it should be.
+		return nil, fmt.Errorf("manifest digest: %s does not match Docker-Content-Digest: %s", digest, checksum)
+	}
+
+	r.manifest = manifest
+	return r.manifest, nil
+}
+
+func (r *remoteImage) RawConfigFile() ([]byte, error) {
+	r.configLock.Lock()
+	defer r.configLock.Unlock()
+	if r.config != nil {
+		return r.config, nil
+	}
+
+	m, err := partial.Manifest(r)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := r.Blob(m.Config.Digest)
+	if err != nil {
+		return nil, err
+	}
+	defer body.Close()
+
+	r.config, err = ioutil.ReadAll(body)
+	if err != nil {
+		return nil, err
+	}
+	return r.config, nil
+}
+
+// TODO(jonjohnsonjr): Wrap the reader in one that performs content validation on EOF.
+func (r *remoteImage) Blob(h v1.Hash) (io.ReadCloser, error) {
+	u := r.url("blobs", h.String())
+	resp, err := r.client.Get(u.String())
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		// TODO(jonjohnsonjr): Parse this into a structured error.
+		resp.Body.Close()
+		return nil, fmt.Errorf("unrecognized status code during blob (%s) GET: %v", h, resp.Status)
+	}
+
+	return resp.Body, nil
```

Comment:
TODO: wrap the `Body` in a `io.ReadCloser` that verifies the digest before returning `io.EOF`?

Reply:
Done